### PR TITLE
BUG: fixes datetime resampling

### LINKED ISSUE #16216: BUG: integer overflow with scipy.resample
### Describe your issue.

I'm working with resampling for my signals and I also need datetime to get resampled which it's easy with the  `t` parameter. 
but the problem is if the `Nx` and `num` are too large (it my case I want to resample the data by *10), we get the overflow problem. The datetime at sometime doesn't produce correct output. I could resolve the issue by writing my own `t` and adds it as a new column to the resampled signal. 
It doesn't produce any error but by checking the outputs you can see the results are not  correct. You can test it with the example code below:

### Reproducing Code Example

```python
from scipy.signal import resample
import pandas as pd
import numpy as np
datetime_test =  pd.date_range(start="2022-02-16 13:51:40.000000",end="2022-02-17 11:41:20.203125",periods =5029134 )

#DatetimeIndex([       '2022-02-16 13:51:40', '2022-02-16 13:51:40.015625',
 #              '2022-02-16 13:51:40.031250', '2022-02-16 13:51:40.046875',
 #              '2022-02-16 13:51:40.062500', '2022-02-16 13:51:40.078125',
 #              '2022-02-16 13:51:40.093750', '2022-02-16 13:51:40.109375',
 #              '2022-02-16 13:51:40.125000', '2022-02-16 13:51:40.140625',
  #             ...
#               '2022-02-17 11:41:20.062500', '2022-02-17 11:41:20.078125',
 #              '2022-02-17 11:41:20.093750', '2022-02-17 11:41:20.109375',
 #              '2022-02-17 11:41:20.125000', '2022-02-17 11:41:20.140625',
 #              '2022-02-17 11:41:20.156250', '2022-02-17 11:41:20.171875',
 #              '2022-02-17 11:41:20.187500', '2022-02-17 11:41:20.203125'],
#              dtype='datetime64[ns]', length=5029134, freq=None)

Nx = len(datetime_test)
num = Nx*10

## scipy:
## new_t = np.arange(0, num) * (t[1] - t[0]) * Nx / float(num) + t[0] 
new_t_scipytest = np.arange(0, num) * (datetime_test[1] - datetime_test[0]) * Nx / float(num) + datetime_test[0]

# array(['2022-02-16T13:51:40.000000000', '2022-02-16T13:51:40.001562500',
 #      '2022-02-16T13:51:40.003125000', ...,
#       '2022-02-16T13:53:05.522297002', '2022-02-16T13:53:05.523859502',
#       '2022-02-16T13:53:05.525422002'], dtype='datetime64[ns]')



## correct verison
## new_t = np.arange(0, num) * (t[1] - t[0])  * (Nx/float(num)) + t[0]
new_t_mytest = np.arange(0, num) * (datetime_test[1] - datetime_test[0]) * (Nx / float(num)) + datetime_test[0]

# array(['2022-02-16T13:51:40.000000000', '2022-02-16T13:51:40.001562500',
#       '2022-02-16T13:51:40.003125000', ...,
#       '2022-02-17T11:41:20.214062500', '2022-02-17T11:41:20.215625000',
#       '2022-02-17T11:41:20.217187500'], dtype='datetime64[ns]')

```

### Error message

```shell
non
```

### SciPy/NumPy/Python version information

1.7.1

### PR DESCRIPTION
**Pull request by SahebehDadboud**
#### Reference issue
Closes #16216

#### What does this implement/fix?
<avoid integer overflow by resampling datetime-->

#### Additional information
<None-->

### CODE REVIEW

**Review by tupui — COMMENTED** [review]
Thank you for the PR @SahebehDadboud. LGTM based on the example in the issue. Could you add a test for this actually? It would just be running your example and checking the values are as expected.

**Comment by SahebehDadboud** [comment]
> Thank you for the PR @SahebehDadboud. LGTM based on the example in the issue. Could you add a test for this actually? It would just be running your example and checking the values are as expected.

working on it.

**Comment by tupui** [comment]
FYI we are freezing the next release at the end of the week.

**Comment by SahebehDadboud** [comment]
> we are freezing the next release at the end of the week.

thanks for the update. I had problems writing a test (as a total beginner). I got sense of it, but still not completely done. I can push with what I wrote already. The test functions works, but testing it with numpy.testing.assert_allclose or any other testing, it gives mismatch or even errors.

**Comment by tupui** [comment]
Feel free to push. It's then easier to help. If you are worried about using too much CI, you can disable some. E.g. add the following in the commit message to not build doc (which here makes sense) `[skip circle]`.

File: scipy/signal/tests/test.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import numpy as np
+import pandas as pd
```
Comment (tupui) [inline scipy/signal/tests/test.py:2]: I am afraid we cannot use Pandas in our tests. Is there a way you could rewrite it without?

File: scipy/signal/tests/test.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+from scipy.signal import resample
+
+
+def test_resample_datetime():
+    n = 5029134
+    signal_test = np.random.uniform(-1841.9731136223052,
```
Comment (tupui) [inline scipy/signal/tests/test.py:8]: We need to use a generator
```python
rng = np.random.default_rng(2985402430427174450683967441442507681)
rng.uniform(...)
```

**Review by tupui — CHANGES_REQUESTED** [review]

File: scipy/signal/tests/test.py (hunk 1 above)
Comment (SahebehDadboud) [reply scipy/signal/tests/test.py:2]: > I am afraid we cannot use Pandas in our tests. Is there a way you could rewrite it without?

can you use datetime?

Comment (tupui) [reply scipy/signal/tests/test.py:2]: Yes of course it's in the standard library. What we cannot do is use libraries we don't have in our testing dependencies.

Comment (SahebehDadboud) [reply scipy/signal/tests/test.py:2]: > Yes of course it's in the standard library. What we cannot do is use libraries we don't have in our testing dependencies.

don't even run the test I wrote. I have no idea why I pushed it. I'm working on a new one with the changes you asked for.

Comment (SahebehDadboud) [reply scipy/signal/tests/test.py:2]: I think it's ready :D

Comment (tupui) [reply scipy/signal/tests/test.py:2]: Thank you, this file needs to be removed though.

Comment (SahebehDadboud) [reply scipy/signal/tests/test.py:2]: please check the one I wrote in test_signaltools.py. test.py is not correct.

**Review by tupui — CHANGES_REQUESTED** [review]

File: scipy/signal/tests/test_resample.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import numpy as np
```
Comment (tupui) [inline scipy/signal/tests/test_resample.py:1]: This file needs to be removed. We already have that in `test_signaltools.py`

**Comment by tupui** [comment]
CI failure is related here.

File: scipy/signal/tests/test_signaltools.py [hunk 4]
```diff
@@ -1240,6 +1241,7 @@ def test_basic(self):
         signal.resample(sig2, num, axis=-1, window=win)
         assert_(win.shape == (160,))
 
+
```
Comment (j-bowhay) [inline scipy/signal/tests/test_signaltools.py:1244]:
```suggestion
```

**Comment by SahebehDadboud** [comment]
> CI failure is related here. Could you also merge main to your branch? Otherwise more CI will fail.

@tupui I've merged the main to my branch. Anything else I have to do?
btw, I've tested the function I wrote and It throws errors. You can of course see the errors.

**Comment by tupui** [comment]
> @tupui I've merged the main to my branch. Anything else I have to do?
> 
> btw, I've tested the function I wrote and It throws errors. You can of course see the errors. 

The CI needs to pass so we need to fix the errors.

**Comment by SahebehDadboud** [comment]
> > @tupui I've merged the main to my branch. Anything else I have to do?
> > btw, I've tested the function I wrote and It throws errors. You can of course see the errors.
> 
> The CI needs to pass so we need to fix the errors.

@tupui I just familiarised myself more with testing. yet on my local machine all of the tests fail. is there anything I can do?

**Review by tupui — CHANGES_REQUESTED** [review]
> @tupui I just familiarised myself more with testing. yet on my local machine all of the tests fail. is there anything I can do?

How did you build SciPy? You need to follow this guide https://scipy.github.io/devdocs/dev/dev_quickstart.html (use conda and dev.py).

Does the test you added work on your side?

Otherwise, please revert the changes in `_lib` and `optimize`. This is making the CI fail.

**Comment by SahebehDadboud** [comment]
> > @tupui I just familiarised myself more with testing. yet on my local machine all of the tests fail. is there anything I can do?
> 
> How did you build SciPy? You need to follow this guide https://scipy.github.io/devdocs/dev/dev_quickstart.html (use conda and dev.py).
> 
> Does the test you added work on your side?
> 
> Otherwise, please revert the changes in `_lib` and `optimize`. This is making the CI fail.

yes I did everything on a Conda env (scipy-dev), but I don't know why nothing works. I might have changed interpreter and now I don't know on which one I built everything. I guess I do everything from scratch. 

that's the last error I got by running tests:
`ImportError: The `scipy` install you are using seems to be broken, (extension modules cannot be imported), please try reinstalling`.

**Comment by SahebehDadboud** [comment]
> > > @tupui I just familiarised myself more with testing. yet on my local machine all of the tests fail. is there anything I can do?
> > 
> > 
> > How did you build SciPy? You need to follow this guide https://scipy.github.io/devdocs/dev/dev_quickstart.html (use conda and dev.py).
> > Does the test you added work on your side?
> > Otherwise, please revert the changes in `_lib` and `optimize`. This is making the CI fail.
> 
> yes I did everything on a Conda env (scipy-dev), but I don't know why nothing works. I might have changed interpreter and now I don't know on which one I built everything. I guess I do everything from scratch.
> 
> that's the last error I got by running tests: `ImportError: The `scipy` install you are using seems to be broken, (extension modules cannot be imported), please try reinstalling`.



> > @tupui I just familiarised myself more with testing. yet on my local machine all of the tests fail. is there anything I can do?
> 
> How did you build SciPy? You need to follow this guide https://scipy.github.io/devdocs/dev/dev_quickstart.html (use conda and dev.py).
> 
> Does the test you added work on your side?
> 
> Otherwise, please revert the changes in `_lib` and `optimize`. This is making the CI fail.

@tupui I was constantly getting error even with building the setups. I'll do everything from scratch

**Comment by tupui** [comment]
What do you mean with building the setups? You must not build all setups yourself. The guide is telling you to

0. Init git submodules
1. Setup a conda env using our environment file
2. Build using `python dev.py build`

If you do anything else, you will likely trouble to make it work or it will just be broken if you miss some manual steps.

**Comment by SahebehDadboud** [comment]
> Looks like we have a bug with GitHub here. I have no clue what happened but it's saying you merge your branch with scipy/main which is not the case since you don't have rights

well I never thought it's that easy to merge sth to main. I hope it didn't make that much troubles. 

regarding `python dev.py build`. it threw errors multiple times. I did the same procedure as you mentioned.

**Comment by tupui** [comment]
No worries, you did not merge anything to `main`. It's just here that it wrongly tells us that. I am not sure what you did to get the UI to do that but as it is your PR is closed and I cannot even reopen it. So you might have to create a new one.

What errors do you have? Can you post a log?

**Comment by SahebehDadboud** [comment]
> No worries, you did not merge anything to `main`. It's just here that it wrongly tells us that. I am not sure what you did to get the UI to do that but as it is your PR is closed and I cannot even reopen it. So you might have to create a new one.
> 
> What errors do you have? Can you post a log?
@tupui  ok I'm losing my brain over this. after doing the steps you mentioned, this is what I get,  and yes the whole time. any help would be appreciated. 

[import scipy
Traceback (most recent call last):
  File "/Users/sahebedadboud/Projects/scipy/scipy/__init__.py", line 121, in <module>
    from scipy.__config__ import show as show_config
ModuleNotFoundError: No module named 'scipy.__config__'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/sahebedadboud/Projects/scipy/scipy/__init__.py", line 126, in <module>
    raise ImportError(msg) from e
ImportError: Error importing SciPy: you cannot import SciPy while
        being in scipy source directory; please exit the SciPy source
        tree first and relaunch your Python interpreter.](url)

**Comment by tupui** [comment]
@SahebehDadboud Did you follow the instructions from the error message?

```
ImportError: Error importing SciPy: you cannot import SciPy while
        being in scipy source directory; please exit the SciPy source
        tree first and relaunch your Python interpreter.
```

Note that if you do this you also need to manually set your PYTHONPATH to the build directory (not the source directory).

Instead, simply use `python dev.py test --help`. You also have other commands such as `shell` and `ipython`.

**Comment by SahebehDadboud** [comment]
> @SahebehDadboud Did you follow the instructions from the error message?
> 
> ```
> ImportError: Error importing SciPy: you cannot import SciPy while
>         being in scipy source directory; please exit the SciPy source
>         tree first and relaunch your Python interpreter.
> ```
> 
> Note that if you do this you also need to manually set your PYTHONPATH to the build directory (not the source directory).
> 
> Instead, simply use `python dev.py test --help`. You also have other commands such as `shell` and `ipython`.

to be honest it worked this time.

**Comment by SahebehDadboud** [comment]
> > @SahebehDadboud Did you follow the instructions from the error message?
> > ```
> > ImportError: Error importing SciPy: you cannot import SciPy while
> >         being in scipy source directory; please exit the SciPy source
> >         tree first and relaunch your Python interpreter.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Note that if you do this you also need to manually set your PYTHONPATH to the build directory (not the source directory).
> > Instead, simply use `python dev.py test --help`. You also have other commands such as `shell` and `ipython`.
> 
> @tupui  to be honest it worked this time.

**Pull request merged**